### PR TITLE
POR-1923 setup dedicated revision rollback endpoint

### PR DESCRIPTION
**Pull request by ianedwards**
## What does this PR do?

- Exposes /apps/{app_name}/rollback, which rolls back to the previous successful revision or the app revision with a given ID if provided
- Changes CLI to call this endpoint

### CODE REVIEW

File: api/server/handlers/porter_app/rollback_revision.go [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+	ctx, span := telemetry.NewSpan(r.Context(), "serve-rollback-app-revision")
+	defer span.End()
+
+	project, _ := ctx.Value(types.ProjectScope).(*models.Project)
+	cluster, _ := ctx.Value(types.ClusterScope).(*models.Cluster)
+
+	telemetry.WithAttributes(span,
+		telemetry.AttributeKV{Key: "project-id", Value: project.ID},
+		telemetry.AttributeKV{Key: "cluster-id", Value: cluster.ID},
+	)
```
Comment (ferozemohideen) [inline api/server/handlers/porter_app/rollback_revision.go:58]: do we need this?

File: api/server/handlers/porter_app/rollback_revision.go [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+	revisions := latestAppRevisionsResp.Msg.AppRevisions
+	if len(revisions) == 0 {
+		return proto, telemetry.Error(ctx, span, nil, "no revisions found for app")
+	}
+
+	skip := 0
+	if revisions[0].RevisionNumber != 0 {
+		skip = 1
+	}
```
Comment (ferozemohideen) [inline api/server/handlers/porter_app/rollback_revision.go:271]: im confused by the skip stuff can you add comments?

File: cli/cmd/v2/rollback.go [hunk 3]
```diff
@@ -30,38 +27,13 @@ func Rollback(ctx context.Context, inp RollbackInput) error {
...
-		return fmt.Errorf("error calling apply endpoint: %w", err)
-	}
-
-	if applyResp.CLIAction != porterv1.EnumCLIAction_ENUM_CLI_ACTION_NONE {
-		return fmt.Errorf("unexpected CLI action: %s", applyResp.CLIAction)
+		return fmt.Errorf("error calling rollback revision endpoint: %w", err)
 	}
 
-	color.New(color.FgGreen).Printf("Successfully rolled back to revision %d\n", rollbackTarget.RevisionNumber) // nolint:errcheck,gosec
+	color.New(color.FgGreen).Printf("Successfully applied revision %s\n", rollbackResp.AppRevisionID) // nolint:errcheck,gosec
```
Comment (ferozemohideen) [inline cli/cmd/v2/rollback.go:37]: does it make sense to also/instead return the revision number?

**Review by ferozemohideen — APPROVED** [review]

File: api/server/handlers/porter_app/rollback_revision.go (hunk 1 above)
Comment (stefanmcshane) [reply api/server/handlers/porter_app/rollback_revision.go:63]: not anymore on porter, but yes on CCP

File: api/server/handlers/porter_app/rollback_revision.go [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+			projectID:     int64(project.ID),
+			appRevisionID: request.AppRevisionID,
+			ccpClient:     c.Config().ClusterControlPlaneClient,
+		})
+		if err != nil {
+			err = telemetry.Error(ctx, span, err, "error getting revision proto")
+			c.HandleAPIError(w, r, apierrors.NewErrPassThroughToClient(err, http.StatusInternalServerError))
+			return
+		}
+	} else {
```
Comment (stefanmcshane) [inline api/server/handlers/porter_app/rollback_revision.go:122]: is this else actually needed? Why not get the last deployed proto, and if that returns as empty get the revision proto?

File: api/server/handlers/porter_app/rollback_revision.go [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+	})
+}
+
+type getRevisionProtoInput struct {
+	projectID     int64
+	appRevisionID string
+	ccpClient     porterv1connect.ClusterControlPlaneServiceClient
+}
+
+func getRevisionProto(ctx context.Context, inp getRevisionProtoInput) (*porterv1.PorterApp, error) {
```
Comment (stefanmcshane) [inline api/server/handlers/porter_app/rollback_revision.go]: To paraphrase Rob Pike's words, which I agree with - if you need words like "get", "do" etc in your function name, then you need a better function name.

File: api/server/handlers/porter_app/rollback_revision.go (hunk 2 above)
Comment (stefanmcshane) [reply api/server/handlers/porter_app/rollback_revision.go:278]: Agreed. Alternatively add a comment to the function on the intention of the code, and change `skip` to something meaningful

File: cli/cmd/v2/rollback.go (hunk 3 above)
Comment (ianedwards) [reply cli/cmd/v2/rollback.go]: changed to show number instead, agreed that's clearer

File: api/server/handlers/porter_app/rollback_revision.go (hunk 5 above)
Comment (ianedwards) [reply api/server/handlers/porter_app/rollback_revision.go]: updated to be more clear

**Pull request merged**